projects/batch-processing/main.go

```diff
@@ -14,45 +24,243 @@ type Converter struct {
+		if err != nil {
+			log.Printf("error uploading image: %v\n", err)
+			failedRecords = append(failedRecords, []string{record[0]})
+		}
+
+		outputRecords = append(outputRecords, []string{record[0], inputFilename, outputFilename, s3url})
+	}
+
+	// Create a CSV file with the output records
+	log.Println("creating output CSV file ... ")
+	outputsFilepath := fmt.Sprintf("./outputs/%s", *outputFilepath)
+	_, err = CreateCSVFile(outputsFilepath, outputRecords)
+	if err != nil {
+		log.Printf("error creating output csv file: %v\n", err)
+		os.Exit(1)
 	}
+	log.Println("output CSV file created successfully")
 
-	// Do the conversion!
-	err := c.Grayscale(*inputFilepath, *outputFilepath)
+	// Create a CSV file with the failed records
+	if len(failedRecords) == 1 {
+		log.Println("no failed records found")
+		os.Exit(0)
+	}
+
+	log.Println("Creating failed CSV file ... ")
+	failedPath := fmt.Sprintf("./outputs/%s", *failedFilepath)
+	_, err = CreateCSVFile(failedPath, failedRecords)
 	if err != nil {
-		log.Printf("error: %v\n", err)
+		log.Printf("error creating failed csv file: %v\n", err)
+		os.Exit(1)
 	}
 
-	// Log what we did
 	log.Printf("processed: %q to %q\n", *inputFilepath, *outputFilepath)
 }
+
+func ReadCSV(filepath string) (records [][]string, err error) {
+	// Open the file
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Read the CSV file
+	r := csv.NewReader(f)
+	records, err = r.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(records) == 0 {
+		return nil, fmt.Errorf("no records found in the csv file")
+	}
+
+	if len(records[0]) > 1 {
+		return records, fmt.Errorf("more than one column is found in the csv file")
+	}
+
+	return records, nil
+}
+
+func DownloadImage(url string) (string, error) {
+	// Create empty file
+	out, err := os.CreateTemp("", "img-*.jpg")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+	filename := out.Name()
+	// Get the image
+	resp, err := http.Get(url)
+	if err != nil {
+		return filename, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return filename, fmt.Errorf("failed to get the image: %v", resp.Status)
+	}
+
+	// Check if it's an image
+	contentType := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(contentType, "image/") {
+		return filename, fmt.Errorf("invalid image type: %v", contentType)
+	}
+
+	defer resp.Body.Close()
+
+	// Write the image to the file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return filename, err
+	}
+	return filename, nil
+}
+
+func (c *Converter) Grayscale(inputFilepath string, outputFilepath string) error {
+	// Convert the image to grayscale using imagemagick
+	// We are directly calling the convert command
+	_, err := c.cmd([]string{
+		"convert", inputFilepath, "-set", "colorspace", "Gray", outputFilepath,
+	})
+	return err
+}
+
+func UploadImage(filename string) (string, error) {
+	// Get the AWS region and role ARN from the environment
+	region := os.Getenv("AWS_REGION")
```

Comment:
I do not like to see code reading from envvars anywhere other than in the main program, ideally.
Or maybe in a library constructor. 

I would turn this into a struct called something like ImageUploader, I'd make the params you're reading here from the env into member variables and pass them in via constructor. These vars should be read in the main func and validated there.